esphome/components/atm90e32/number/__init__.py

```diff
@@ -0,0 +1,137 @@
+    UNIT_VOLT,
+)
+
+from .. import atm90e32_ns
+from ..sensor import ATM90E32Component
+
+ATM90E32Number = atm90e32_ns.class_(
+    "ATM90E32Number", number.Number, cg.Parented.template(ATM90E32Component)
+)
+
+CONF_REFERENCE_CURRENT = "reference_current"
+PHASE_KEYS = [CONF_PHASE_A, CONF_PHASE_B, CONF_PHASE_C]
+
+
+def reference_voltage_schema():
+    return cv.All(
+        cv.Schema(
+            {
+                cv.Required(CONF_NAME): cv.string,
+                cv.Optional("mode", default="box"): cv.string,
+                cv.Optional("optimistic", default=True): cv.boolean,
+                cv.Optional("min_value", default=100.0): cv.float_,
+                cv.Optional("max_value", default=260.0): cv.float_,
+                cv.Optional("step", default=0.1): cv.float_,
+            }
+        ).extend(
+            number.number_schema(
+                class_=ATM90E32Number,
+                unit_of_measurement=UNIT_VOLT,
+                entity_category=ENTITY_CATEGORY_CONFIG,
+                icon="mdi:power-plug",
+            )
+        )
+    )
+
+
+def reference_current_schema():
+    return cv.All(
+        cv.Schema(
+            {
+                cv.Required(CONF_NAME): cv.string,
+                cv.Optional("mode", default="box"): cv.string,
+                cv.Optional("optimistic", default=True): cv.boolean,
+                cv.Optional("min_value", default=1.0): cv.float_,
+                cv.Optional("max_value", default=200.0): cv.float_,
+                cv.Optional("step", default=0.1): cv.float_,
+            }
+        ).extend(
+            number.number_schema(
+                class_=ATM90E32Number,
+                unit_of_measurement=UNIT_AMPERE,
+                entity_category=ENTITY_CATEGORY_CONFIG,
+                icon="mdi:home-lightning-bolt",
+            )
+        )
+    )
+
+
+REFERENCE_VOLTAGE_SCHEMA = cv.Schema(
+    {
+        cv.Optional(CONF_PHASE_A): reference_voltage_schema(),
+        cv.Optional(CONF_PHASE_B): reference_voltage_schema(),
+        cv.Optional(CONF_PHASE_C): reference_voltage_schema(),
+    }
+)
+
+REFERENCE_CURRENT_SCHEMA = cv.Schema(
+    {
+        cv.Optional(CONF_PHASE_A): reference_current_schema(),
+        cv.Optional(CONF_PHASE_B): reference_current_schema(),
+        cv.Optional(CONF_PHASE_C): reference_current_schema(),
+    }
+)
+
+CONFIG_SCHEMA = cv.Schema(
+    {
+        cv.Required(CONF_ID): cv.use_id(ATM90E32Component),
+        cv.Optional(CONF_REFERENCE_VOLTAGE): REFERENCE_VOLTAGE_SCHEMA,
+        cv.Optional(CONF_REFERENCE_CURRENT): REFERENCE_CURRENT_SCHEMA,
+    }
+)
+
+
+async def to_code(config):
+    parent = await cg.get_variable(config[CONF_ID])
+
+    if CONF_REFERENCE_VOLTAGE in config:
+        voltage_cfg = config[CONF_REFERENCE_VOLTAGE]
+        voltage_objs = [None, None, None]
+
+        for i, key in enumerate(PHASE_KEYS):
+            if key in voltage_cfg:
+                validated = reference_voltage_schema()(voltage_cfg[key])
```

Comment:
Config is already validated before to_code is called
```suggestion
            if validated := voltage_cfg.get(key):
```